crates/ruff_python_parser/src/invalid.rs

```diff
@@ -0,0 +1,810 @@
+/*!
+Defines some helper routines for rejecting invalid Python programs.
+
+These routines are named in a way that supports qualified use. For example,
+`invalid::assignment_targets`.
+*/
+
+use lalrpop_util::ParseError as LalrpopError;
+
+use {ruff_python_ast::Expr, ruff_text_size::TextSize};
+
+use crate::{
+    lexer::{LexicalError, LexicalErrorType},
+    token::Tok,
+};
+
+/// Returns an error for invalid assignment targets.
+///
+/// # Errors
+///
+/// This returns an error when any of the given expressions are themselves
+/// or contain an expression that is invalid on the left hand side of an
+/// assignment. For example, all literal expressions are invalid assignment
+/// targets.
+pub(crate) fn assignment_targets(
+    targets: &[Expr],
+) -> Result<(), LalrpopError<TextSize, Tok, LexicalError>> {
+    for t in targets {
+        assignment_target(t)?;
+    }
+    Ok(())
+}
+
+/// Returns an error if the given target is invalid for the left hand side of
+/// an assignment.
+///
+/// # Errors
+///
+/// This returns an error when the given expression is itself or contains an
+/// expression that is invalid on the left hand side of an assignment. For
+/// example, all literal expressions are invalid assignment targets.
+pub(crate) fn assignment_target(
+    target: &Expr,
+) -> Result<(), LalrpopError<TextSize, Tok, LexicalError>> {
```

Comment:
I played around with it a bit locally because I wasn't sure -- but it looks like these can all just be `Result<(), LexicalError>` without the `LalrpopError` wrapper, if you do something like this:

```diff
diff --git a/crates/ruff_python_parser/src/invalid.rs b/crates/ruff_python_parser/src/invalid.rs
index 99dfc0140..0f53cc144 100644
--- a/crates/ruff_python_parser/src/invalid.rs
+++ b/crates/ruff_python_parser/src/invalid.rs
@@ -22,9 +22,7 @@ use crate::{
 /// or contain an expression that is invalid on the left hand side of an
 /// assignment. For example, all literal expressions are invalid assignment
 /// targets.
-pub(crate) fn assignment_targets(
-    targets: &[Expr],
-) -> Result<(), LalrpopError<TextSize, Tok, LexicalError>> {
+pub(crate) fn assignment_targets(targets: &[Expr]) -> Result<(), LexicalError> {
     for t in targets {
         assignment_target(t)?;
     }
@@ -39,20 +37,17 @@ pub(crate) fn assignment_targets(
 /// This returns an error when the given expression is itself or contains an
 /// expression that is invalid on the left hand side of an assignment. For
 /// example, all literal expressions are invalid assignment targets.
-pub(crate) fn assignment_target(
-    target: &Expr,
-) -> Result<(), LalrpopError<TextSize, Tok, LexicalError>> {
+pub(crate) fn assignment_target(target: &Expr) -> Result<(), LexicalError> {
     // Allowing a glob import here because of its limited scope.
     #[allow(clippy::enum_glob_use)]
     use self::Expr::*;
 
-    let err = |location: TextSize| -> LalrpopError<TextSize, Tok, LexicalError> {
+    let err = |location: TextSize| -> LexicalError {
         let msg = "invalid assignment target";
-        let error = LexicalError {
+        LexicalError {
             error: LexicalErrorType::OtherError(msg.to_string()),
             location,
-        };
-        LalrpopError::User { error }
+        }
     };
     match *target {
         BoolOp(ref e) => return Err(err(e.range.start())),
@@ -84,24 +79,23 @@ pub(crate) fn assignment_target(
         EllipsisLiteral(ref e) => return Err(err(e.range.start())),
         // The only nested expressions allowed as an assignment target
         // are star exprs, lists and tuples.
-        Starred(ref e) => assignment_target(&e.value)?,
-        List(ref e) => assignment_targets(&e.elts)?,
-        Tuple(ref e) => assignment_targets(&e.elts)?,
+        Starred(ref e) => assignment_target(&e.value),
+        List(ref e) => assignment_targets(&e.elts),
+        Tuple(ref e) => assignment_targets(&e.elts),
         // Subscript is recursive and can be invalid, but aren't syntax errors.
         // For example, `5[1] = 42` is a type error.
-        Subscript(_) => {}
+        Subscript(_) => Ok(()),
         // Similar to Subscript, e.g., `5[1:2] = [42]` is a type error.
-        Slice(_) => {}
+        Slice(_) => Ok(()),
         // Similar to Subscript, e.g., `"foo".y = 42` is an attribute error.
-        Attribute(_) => {}
+        Attribute(_) => Ok(()),
         // These are always valid as assignment targets.
-        Name(_) => {}
+        Name(_) => Ok(()),
         // This isn't in the Python grammar but is Jupyter notebook specific.
         // It seems like this should be an error, but I'm not certain about it
         // at atime of writing. So we allow it. ---AG
-        IpyEscapeCommand(_) => {}
+        IpyEscapeCommand(_) => Ok(()),
     }
-    Ok(())
 }
```